benchmarks/ref/include/Bench.h

```diff
@@ -0,0 +1,91 @@
+//===- Bench.h - ----------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "Perf.h"
+#include "Tensor.h"
+#include <vector>
+
+// KernelInterface: Must be implemented by kernels to use Benchmark class.
+//
+// T is a numeric type (int, float, double), and represents the payload
+// type of the input and output tensors, not necessarily the accumulation
+// and working internal tensors.
+template <typename T> struct KernelInterface {
+  virtual ~KernelInterface() {}
+  // Runs the reference kernel with the list of arguments.
+  // Output needs to be passed in the args list and is up to the user to
+  // implement the correct logic.
+  virtual void runRef(std::vector<T> &args) = 0;
+};
+
+// Benchmark: runs a kernel multiple times, takes timings, return avg.
+//
+// Supports warm-up (JIT compilers, cache flush, etc), tensor initialization,
+// performance tracking, output comparison, etc.
+//
+// Usage:
+//  * Implement a kernel that implements the KernelInterface
+//  * Define a variable as Benchmark<MyKernel>
+template <class Kernel, class T> class Benchmark {
```

Comment:
We had this before, but we moved to IR based timer, right?

This should be good for both CPU and GPU.

Reply:
That's for my CUDA baseline implementations.